### PR TITLE
[56339] Split content of Admin/Design page into separate tabs

### PR DESCRIPTION
**Pull request by bsatarnejad**
https://community.openproject.org/work_packages/56339/activity

### CODE REVIEW

File: spec/controllers/custom_styles_controller_spec.rb [hunk 1]
```diff
@@ -85,7 +85,8 @@
         let(:valid) { true }
 
         it "redirects to show" do
-          expect(response).to redirect_to action: :show
+          expect(response).to be_redirect
```
Comment (HDinger) [inline spec/controllers/custom_styles_controller_spec.rb]: You are now only testing whether there is a redirect but not to which page. I guess it makes sense to check that as well as it could also be an error page and the test would be green

File: app/controllers/custom_styles_controller.rb [hunk 2]
```diff
@@ -52,10 +52,22 @@ def show
...
+  def branding
+    @custom_style = CustomStyle.current || CustomStyle.new
+    @current_theme = @custom_style.theme
+    @theme_options = options_for_theme_select
+  end
+
+  def pdf_export_styles
+    @custom_style = CustomStyle.current || CustomStyle.new
+    @current_theme = @custom_style.theme
+    @theme_options = options_for_theme_select
```
Comment (HDinger) [inline app/controllers/custom_styles_controller.rb]: This block is repeated three times in total in this controller.

**Review by HDinger — CHANGES_REQUESTED** [review]
It works as expected. I have however some remark with regards to the code structure. You are currently mixing and repeating things a bit. 
When you look at the code structure of other admin pages, we usually have one file per base route (`show`, `new`, `index`, etc). Addtional stuff is extracted in partials. That also requires separate controllers for each module. 

You have a `show` page, and a `branding` and `pdf_export_styles` page while all three of them are routed with the same controller. The controller actions are however exact duplicates. So there  is no really benefit of doing it like that. This approach also has another drawback: In each file, you have to repeat the call of the header with a magic number passed through:

```
<%=
  render(Admin::DesignHeaderComponent.new(
    title: t(:"admin.custom_styles.tab_branding"),
    selected: 2,
    ))
%>
```

As soon as we add another tab in between, the selection will not work any more and has to be changes not only in the view files but also in the header component as well. 

I am aware that you basically copied that header approach from attachments, but I think it is a bad pattern. Further, attachements has extracted stuff in individual controllers since the logic is very different.

Tl;dr; Instead of extracting all in individual files and repeating stuff, please use partials. You then only have to call the header once  in the `show` file and can probably even put the repeated theme selector in there. The partials would then only contain the actual differences.

File: app/components/admin/design_header_component.html.erb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                             t(:label_custom_style)])
+    header.with_description { t(:label_custom_style_description) }
+    header.with_tab_nav(label: nil) do |tab_nav|
+      @tabs.each do |tab|
+        tab_nav.with_tab(selected: selected_tab(@tabs) == tab, href: tab[:path]) do |t|
+          t.with_text { tab[:label] }
+        end
+      end
+    end if @tabs.present?
+  end
```
Comment (EinLama) [inline app/components/admin/design_header_component.html.erb]: Nitpick: I'd suggest to use an explicit if-block instead of a trailing-if when your happy execution-path spans over more than one line. It makes the body of the if-clause much more readable. Also, I prefer to give the code at least some breathing room by adding an empty line before a bigger code chunk. Combined, my suggestion would look like this: what do you think?

```suggestion
  render(Primer::OpenProject::PageHeader.new) do |header|
    header.with_title { t(:label_custom_style) }
    header.with_breadcrumbs([{ href: admin_index_path, text: t(:label_administration) },
                             t(:label_custom_style)])
    header.with_description { t(:label_custom_style_description) }
    
    if @tabs.present?
      header.with_tab_nav(label: nil) do |tab_nav|
        @tabs.each do |tab|
          tab_nav.with_tab(selected: selected_tab(@tabs) == tab, href: tab[:path]) do |t|
            t.with_text { tab[:label] }
          end
        end
      end
    end
  end
```

File: app/components/admin/design_header_component.rb [hunk 4]
```diff
@@ -26,38 +28,15 @@
...
-    end
+module Admin
+  class DesignHeaderComponent < ApplicationComponent
+    include OpPrimer::ComponentHelpers
+    include ApplicationHelper
+    include TabsHelper
 
-    it "shows a replace button" do
-      expect(rendered).to include "Replace"
+    def initialize(tabs: nil)
```
Comment (EinLama) [inline app/components/admin/design_header_component.rb]: Whenever you expect a nullable value, consider using the "blank" alternative to `nil` first.

Here, we expect a list of tabs, but there could be no tabs. Instead of expressing this as `nil`, you could also express this as an empty list. This would help the reader understand what type to expect here. Like so:

```suggestion
    def initialize(tabs: [])
```

Luckily, Rails `present?` method will return `false` for empty Arrays (as well as empty Strings and empty Hashes), so this code change will work the same as `nil` for your purposes.

File: app/controllers/custom_styles_controller.rb [hunk 5]
```diff
@@ -44,10 +44,16 @@ class CustomStylesController < ApplicationController
                      only: UNGUARDED_ACTIONS
   no_authorization_required! *UNGUARDED_ACTIONS
 
+  def default_url_options
+    super.merge(tab: params[:tab])
+  end
```
Comment (EinLama) [inline app/controllers/custom_styles_controller.rb:49]: Please ensure an [empty line between methods](https://github.com/rubocop/ruby-style-guide?tab=readme-ov-file#empty-lines-between-methods).

File: app/controllers/custom_styles_controller.rb [hunk 6]
```diff
@@ -44,10 +44,16 @@ class CustomStylesController < ApplicationController
...
   def show
     @custom_style = CustomStyle.current || CustomStyle.new
     @current_theme = @custom_style.theme
     @theme_options = options_for_theme_select
+    if params[:tab].blank?
```
Comment (EinLama) [inline app/controllers/custom_styles_controller.rb:56]: Nitpick: Consider adding an empty line before the if-statement to give the condition-part of the method a visual distinction to the assignment-part of the method.

I know that this controller is rather crammed in the other methods, but it would really improve the readability if we stepped up our formatting here.

File: app/controllers/custom_styles_controller.rb [hunk 7]
```diff
@@ -148,12 +154,7 @@ def update_themes
     call.on_failure do
       flash[:error] = call.message
     end
-
-    redirect_to action: :show
-  end
-
-  def show_local_breadcrumb
-    false
+    redirect_to custom_style_path
```
Comment (EinLama) [inline app/controllers/custom_styles_controller.rb:158]: Nitpick: add a blank line above this to improve readability and consistency.

File: app/views/custom_styles/_interface.html.erb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  <%# Fill in the computed css variables in the inputs as their defaults %>
+  <%= nonced_javascript_tag do %>
+    var computedStyle = getComputedStyle(document.documentElement);
+    document
+    .querySelectorAll('.design-color--variable-input')
+    .forEach(function(el) {
+    if (!el.value || el.value === '') {
+    el.placeholder = computedStyle.getPropertyValue('--' + el.dataset.variableName).trim();
+    }
+    });
```
Comment (EinLama) [inline app/views/custom_styles/_interface.html.erb]: Nitpick: This small snippet deserves some indentation to make it more readable.

```suggestion
    var computedStyle = getComputedStyle(document.documentElement);
    
    document
        .querySelectorAll('.design-color--variable-input')
        .forEach(function(el) {
            if (!el.value || el.value === '') {
                el.placeholder = computedStyle.getPropertyValue('--' + el.dataset.variableName).trim();
            }
        });
```

It would be possible to add a slim Stimulus controller that takes care of setting the properties of these fields. Not sure if we usually do it for small tasks like this or not. If you did consider Stimulus here and decided against it, ignore this idea. If you did not, feel free to consider it :)

File: app/views/custom_styles/_interface.html.erb [hunk 9]
```diff
@@ -0,0 +1,45 @@
+<%= form_tag update_design_colors_path, method: :post, class: "form" do %>
+  <section class="form--section">
+    <fieldset class="form--fieldset">
+      <legend class="form--fieldset-legend"><%= I18n.t(:label_interface_colours) %></legend>
+      <p> <%= t('label_interface_colours_description') %> </p>
```
Comment (EinLama) [inline app/views/custom_styles/_interface.html.erb]: This here should be American English and therefore `label_interface_colors_description`. The current translation for this line is British, too. It also should be americanized to stay consistent. I've added a suggestion in the corresponding YAML file :) 

EDIT: Ah, I can see that the design files mistakenly used the British version. This is probably why you added them as-is. We still have to correct them here to keep our localization consistent.

File: app/views/custom_styles/show.html.erb [hunk 10]
```diff
@@ -27,18 +27,14 @@ See COPYRIGHT and LICENSE files for more details.
 
 ++#%>
 
-<% html_title t(:label_administration), t(:label_custom_style) %>
-
+<% tabs = design_tabs %>
```
Comment (EinLama) [inline app/views/custom_styles/show.html.erb]: This variable is never used and can be removed.

File: app/views/custom_styles/show.html.erb [hunk 11]
```diff
@@ -27,18 +27,14 @@ See COPYRIGHT and LICENSE files for more details.
...
 <%=
-  render(Primer::OpenProject::PageHeader.new) do |header|
-    header.with_title { t(:label_custom_style) }
-    header.with_breadcrumbs([{ href: admin_index_path, text: t(:label_administration) },
-                             t(:label_custom_style)])
-  end
+  render(Admin::DesignHeaderComponent.new(
+    tabs: design_tabs,
+    ))
```
Comment (EinLama) [inline app/views/custom_styles/show.html.erb]: Nitpick: This could be one line:

```ruby
<%= render Admin::DesignHeaderComponent.new(tabs: design_tabs) %>
```

File: config/locales/en.yml [hunk 12]
```diff
@@ -2044,7 +2047,8 @@ en:
   label_additional_workflow_transitions_for_assignee: "Additional transitions allowed when the user is the assignee"
   label_additional_workflow_transitions_for_author: "Additional transitions allowed when the user is the author"
   label_administration: "Administration"
-  label_advanced_settings: "Advanced settings"
+  label_interface_colours: "Interface colours"
+  label_interface_colours_description: "These colours control how the application looks. If you modify them the theme will automatically be changed to Custom theme, but we can’t assure the compliance of the accessibility contrast minimums (WCAG 2.1). "
```
Comment (EinLama) [inline config/locales/en.yml]: colours -> colors

```suggestion
  label_interface_colors: "Interface colors"
  label_interface_colors_description: "These colors control how the application looks. If you modify them the theme will automatically be changed to Custom theme, but we can’t assure the compliance of the accessibility contrast minimums (WCAG 2.1). "
```

File: config/locales/en.yml [hunk 13]
```diff
@@ -2158,6 +2162,7 @@ en:
   label_custom_field_plural: "Custom fields"
   label_custom_field_default_type: "Empty type"
   label_custom_style: "Design"
+  label_custom_style_description: "Choose how OpenProject looks to you with themes, select your default colours to use in the app and how exports look like."
```
Comment (EinLama) [inline config/locales/en.yml]: colours -> colors
```suggestion
  label_custom_style_description: "Choose how OpenProject looks to you with themes, select your default colors to use in the app and how exports look like."
```

File: spec/controllers/custom_styles_controller_spec.rb (hunk 1 above)
Comment (EinLama) [reply spec/controllers/custom_styles_controller_spec.rb]: This is still true, also for all of the following test cases. Are there any issues with doing this? 🤔

File: spec/features/custom_styles/tabs_spec.rb [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require "spec_helper"
+
+RSpec.describe "interface tab" do
```
Comment (EinLama) [inline spec/features/custom_styles/tabs_spec.rb]: This naming is a bit misleading. This spec does far more than only testing the interface tab. It starts there, but it will ensure the entire design section works. I suggest to rename this spec, also consider renaming the file to show it does many nice things 👍🏻

**Review by EinLama — CHANGES_REQUESTED** [review]
Looks nice and does what it should, nicely done! 👍🏻 

I've found one issue with the **localization**. It uses British English in one spot where it should be **US English**. Another issue regards the **naming of a spec** and the **redirect-verification** within the specs.

The other comments and suggestions are mostly code style related nitpicks and nothing functional. Please consider them and add the ones that you agree with.

File: app/controllers/custom_styles_controller.rb (hunk 7 above)
Comment (EinLama) [reply app/controllers/custom_styles_controller.rb:158]: About the Rubocop error:

You can either ignore it, or move the theme selection into a private method:
```ruby
  def update_themes
    call = ::Design::UpdateDesignService
      .new(theme_from_params)
      .call

    call.on_success do
      flash[:notice] = I18n.t(:notice_successful_update)
    end

    call.on_failure do
      flash[:error] = call.message
    end

    redirect_to custom_style_path
  end

  private

  def theme_from_params
    OpenProject::CustomStyles::ColorThemes.themes.find { |t| t[:theme] == params[:theme] }
  end
```

File: app/components/admin/design_header_component.html.erb (hunk 3 above)
Comment (bsatarnejad) [reply app/components/admin/design_header_component.html.erb]: yeah, your suggestion is more readable 👍

**Review by EinLama — APPROVED** [review]
LGTM 🎉 great work!
